interface/src/Application.h

```diff
@@ -505,7 +510,7 @@ private slots:
 
     void notifyPacketVersionMismatch();
 
-    void loadSettings();
+    void loadSettings(const QCommandLineParser* parser);
```

Comment:
Why a pointer here and not a reference?

Reply:
Personal prejudice. I read about references once but they just seemed like worse pointers to me. :p

Practically speaking, the only reason I can think of is that you might want to delete the parser to free up memory once the program is initialised. I didn't do that because I wasn't sure if you'd want to. Seems objectively better in the practical sense but maybe a stylistic choice. If you do want to free up the space, it would be good to confirm that the parser isn't being referenced after the application is fully initialised (not just constructed)-- it probably shouldn't be and if it is, that should be trivial to fix. You're welcome to change the pointer to a reference if you prefer though.

Reply:
Sorry for the late reply, got sick, bunch of stuff got out of whack.

A reference is better in this context. A reference can't be a null pointer without invoking undefined behavior, at which point all hope is lost anyway. So if the function is operating under the assumption that `parser` will be set to something useful, a reference is a better fit.

Also, I'm not sure what's the relationship to deletion there. You can't delete something passed with a reference without taking its address -- that's not considered a good style. The general idea is that if you get a reference passed in something else has ownership of the data, and you're not supposed to mess with it.

In any case, deallocation wouldn't be something it'd be appropriate for `loadSettings` to perform.

Reply:
It's ok.

I still don't really understand the value of references (not a pun) but you're welcome to use a reference if that's what you prefer. My pull requests are basically just suggestions. 👍

Reply:
The value is basically that it's a pointer with added restrictions that you want to exist a lot of the time. If you get a pointer it could be `nullptr`, which means now you have to figure out what to do about that. Do you check for it? Could null be a valid value? What if you get one, do you return a special value, do you silently ignore it, do you skip something, do you crash the program? It's potentially troublesome, you may need to reorder code so that any checks/cleanup are correctly ordered, which could get messed up later.

If you're guaranteed to have a valid object as an input it removes all that from consideration.

Reply:
So what you're saying is that passing a reference is the same as `if (parser == nullptr) return;`, just with a different format? I assume it would still have to be tested at run-time.

I don't think it's possible for parser to be nullptr in this case but I could be wrong.

Reply:
No, you can't normally create a reference to nullptr. The compiler just won't allow it. To create a reference you have to take the address of an already existing, valid object.

The one way to do it anyway is to dereference a null pointer, but that is in itself undefined behavior. And at that point you can consider the program irrevocably broken.

Reply:
Right, so the test `if (parser == nullptr)` must occur at run-time regardless of whether a pointer or reference is used. Testing it explicitly is just more flexible, since you could return from the function, run a quit routine, or whatever is appropriate. It's also more clear to read for programmers who aren't very familiar with C++. If using references eliminates the run-time test then I would prefer it. Do you know?

Reply:
There is no test. You can't create a reference to nullptr with valid code.

https://godbolt.org/z/7c6cv8M5M

Try changing it to: `iTakeAReference(nullptr);` -- it'll simply fail to compile.

Reply:
No, I mean the that the reference is probably tested implicitly when the function is called to make sure it is not a nullptr. It might be possible for compilers to discern in some cases that the value could not possibly be nullptr at compile time, I'm not sure if they do. Or maybe pointers can not be passed as references in general?

Reply:
There is no runtime test, no. The language's syntax prevents a reference to a nullptr from being created in the first place. Try playing with the linked program.

Reply:
I am. It seems like pointers can be passed by reference, so the program must check to make sure the pointers are not null. Maybe I'm misunderstanding something. 🤔
https://godbolt.org/z/qxeq8E61Y

Reply:
Oh, you probably mean that it's not possible for the value unless it is a reference to a pointer. That makes sense.

Reply:
I'm still not sure how I feel about references in general but I will try to change this for you since you went to the trouble of explaining references for me. Hopefully nothing gets broken. My ability to test these parameters is limited.

Reply:
Is this how you wanted it? https://github.com/vircadia/vircadia/pull/1428/commits/46d2ff0d0774c0ac7475bc6a40c2da5eac100e94

Reply:
Sorry for jumping in, but I'm a total c++ fanboy.
Yes the point is to replace the pointer to the value with a reference to the value. There are no runtime checks because it's less of an absolute safety guarantee and more of a clarification for the programmers to prevent mistakes. If you have a function that accepts a pointer there is always a question of who's going to be handling null, the implementer or the caller, assuming they are two different people (or same person at different times). The implementer is like "ok do I check it and return an error or something? But what if the caller can guarantee it is not null and doesn't want the overhead of the check?" and the caller is like "ok do I check it? But what if the implementer already checked it and then I'll be checking for no reason". Sometimes it can all work out, other times you end up with either too many or not enough null checks. 
 The reference settles the matter - if the function accepts a reference to a value it will not handle null (as you are not allowed to pass it null, which is a consequence of the language rules, which the compiler will check if/when it can, but otherwise it's up to the programmers to adhere to in general), otherwise if it accepts a pointer it promises to handle null somehow so it's ok to pass it. So as a rule of thumb if your logic requires a value, you accept a reference instead of something like top level null check that just returns, and you use a pointer if the parameter is optional and checking it is part of your logic.

Reply:
That makes a lot of sense. It feels like a kind of heavy handed way to solve such a small problem but the problem does come up a lot and maybe syntax is a good way to fix it. Could they have added a pointer-specific constraint "nonull" like "const"? That feels more intuitive to me but also takes up a lot more space per line. What do you think?

Reply:
The reference is actually a big part of the language, and it serves a lot of other purposes as well. This is just one of the use cases, and not the sole point of the design. Some other differences from pointers that come to mind:
A reference is completely transparent alias for the value, you can't get a handle on it in any way, it has no address, (if you try to take an address you get the address of the value it refers to), so you can't have a pointer to a reference, or a reference to a reference.
There is a special magic about const references in particular - they can bind to temporaries and extend their lifetime. 
Since c++11 there are rvalue references that allow even greater control over temporaries. 

I don't know the actual history of it, but in general c++ is very big on value semantics and that's a plausible explanation. If you asked a c++ programmer what's would be a perfect world to them, they would likely say "a world where everything behaves like a value", and a lot of the language's design is about achieving that dream. From that point of view references are an attempt to bring value semantics to the world of indirection.

Reply:
Interesting. I wish the one C++ book I read had introduced them that well. :p

I would say that I'm a fan of pointers in general but that's not to say they're perfect. I won't ask any more questions here because I don't want to spam Dale and the Vircadia Discord's GitHub channel too much, but I will plan on looking into references a bit more sometime. Thanks for jumping in.

Reply:
Since Dale doesn't seem to mind, do you know any good books to recommend?

Reply:
It was all courses, talks and articles for me, along with a lot of practice, so can never remember or find something specific to recommend ._. I would say there is a general lack of good beginner materials, so it's normal to just stick to one or two things that you like and then slowly absorb the rest as you go.

Reply:
Ok. I know it's hard to keep books up to date but I am specifically in need of books to read. I don't mind if they're advanced as long as they're well organised, accurate, and complete in their topics, Let me know if you ever run into any. :)

Reply:
Well http://elementsofprogramming.com/eop.pdf is on my must-read list. It's a bit dense and abstract though.

Reply:
Ok, thanks! I'm not very good at math but this seems like a useful perspective on programming. I'll give it a shot. :)

Reply:
It looks like _Embracing Modern C++ Safely_ might be interesting. I will order it as soon as it's available because I like Vittorio Romeo's work on open source projects.

Reply:
This one is kind of expensive but let me know what you think of it once it arrives. If you like it then I will buy a copy too.